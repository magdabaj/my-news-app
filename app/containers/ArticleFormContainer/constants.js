/*
 *
 * ArticleFormContainer constants
 *
 */

export const ADD_ARTICLE = "app/ArticleFormContainer/ADD_ARTICLE";
export const ADD_ARTICLE_SUCCESS = "app/ArticleFormContainer/ADD_ARTICLE_SUCCESS";
export const ADD_ARTICLE_ERROR = "app/ArticleFormContainer/ADD_ARTICLE_ERROR";
export const ADD_ARTICLE_CANCELED = "app/ArticleFormContainer/ADD_ARTICLE_CANCELED";

