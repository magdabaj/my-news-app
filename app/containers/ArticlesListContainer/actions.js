/*
 *
 * ArticlesListContainer actions
 *
 */

import {
  GET_ALL_ARTICLES,
  GET_ARTICLES_FOR_TAG,
  GET_ARTICLES_FOR_TAG_FAILED,
  GET_ARTICLES_FOR_TAG_SUCCESS, START_ADD
} from "./constants";

export const getAllArticles = () => ({
  type: GET_ALL_ARTICLES
});

export const getArticlesForTag = tagId => ({
  type: GET_ARTICLES_FOR_TAG,
  tagId,
});

export const getArticlesForTagSuccess = articles => ({
  type: GET_ARTICLES_FOR_TAG_SUCCESS,
  articles,
});

export const getArticlesForTagFailed = error => ({
  type: GET_ARTICLES_FOR_TAG_FAILED,
  error,
});

export const startAdd = tagName => ({
  type: START_ADD,
  tagName,
});