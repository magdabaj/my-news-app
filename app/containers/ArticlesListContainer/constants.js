/*
 *
 * ArticlesListContainer constants
 *
 */

export const GET_ALL_ARTICLES = "app/ArticlesListContainer/GET_ALL_ARTICLES";

export const GET_ARTICLES_FOR_TAG = "app/ArticlesListContainer/GET_ARTICLES_FOR_TAG";
export const GET_ARTICLES_FOR_TAG_SUCCESS = "app/ArticlesListContainer/GET_ARTICLES_FOR_TAG_SUCCESS";
export const GET_ARTICLES_FOR_TAG_FAILED = "app/ArticlesListContainer/export const GET_ARTICLES_FOR_TAG_FAILED";
