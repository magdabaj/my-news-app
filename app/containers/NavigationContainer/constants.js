/*
 *
 * NavigationContainer constants
 *
 */

export const GET_TAGS = "app/NavigationContainer/GET_TAGS";
export const GET_TAGS_SUCCESS = "app/NavigationContainer/GET_TAGS_SUCCESS";
export const GET_TAGS_FAILED = "app/NavigationContainer/GET_TAGS_FAILED";

export const GET_ARTICLES_BY_TAG = "app/NavigationContainer/GET_ARTICLES_BY_TAG";
