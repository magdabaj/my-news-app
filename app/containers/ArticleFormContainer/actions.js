/*
 *
 * ArticleFormContainer actions
 *
 */

import {ADD_ARTICLE, ADD_ARTICLE_CANCELED, ADD_ARTICLE_ERROR, ADD_ARTICLE_SUCCESS} from "./constants";

export const addArticle = (article, tagId, user) => ({
  type: ADD_ARTICLE,
  article,
  tagId,
  user
});

export const addArticleSuccess = article => ({
  type: ADD_ARTICLE_SUCCESS,
  article,
});

export const addArticleError = error => ({
  type: ADD_ARTICLE_ERROR,
  error,
});

export const addArticleCanceled = () => ({
  type: ADD_ARTICLE_CANCELED,
});