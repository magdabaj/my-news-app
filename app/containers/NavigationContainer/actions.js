/*
 *
 * NavigationContainer actions
 *
 */

import { GET_TAGS, GET_ARTICLES_BY_TAG, GET_TAGS_FAILED, GET_TAGS_SUCCESS } from "./constants";

export const getTags = () => ({
  type: GET_TAGS
});

export const getTagsSuccess = tags => ({
  type: GET_TAGS_SUCCESS,
  tags
});

export const getTagsFailed = error => ({
  type: GET_TAGS_FAILED,
  error
});

export const getArticlesByTag = tagId => ({
  type: GET_ARTICLES_BY_TAG,
  tagId,
});
