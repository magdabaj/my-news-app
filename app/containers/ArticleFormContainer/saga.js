import { take, call, put, select, all, takeLatest } from 'redux-saga/effects';
import {ADD_ARTICLE, ADD_ARTICLE_CANCELED} from "./constants";
import {addArticle} from "../../utils/api/articlesApi";
import {addArticleError, addArticleSuccess} from "./actions";
import {goBack} from "react-router-redux";

export function* addArticleSaga(action) {
    const {article, tagId, userId} = action;
    try {
        const createdArticle = yield call(addArticle, article, tagId, userId);
        yield put(addArticleSuccess(article));
    } catch (e) {
        yield put(addArticleError(e.message))
    }
}

export function* watchArticleAdd() {
    yield takeLatest(ADD_ARTICLE, addArticleSaga);
}

export function* watchAddCanceled() {
    yield takeLatest(ADD_ARTICLE_CANCELED, () => put(goBack()));
}

// Individual exports for testing
export default function* rootSaga() {
  yield all([
      watchArticleAdd(),
          watchAddCanceled()
      ])
}
