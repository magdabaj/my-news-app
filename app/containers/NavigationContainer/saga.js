import { take, call, put, select, takeLatest, all } from 'redux-saga/effects';
import { push } from 'react-router-redux';
import { fetchTags } from '../../utils/api/tagsApi';
import {getTagsFailed, getTagsSuccess} from "./actions";
import {GET_TAGS, GET_TAGS_SUCCESS, SELECT_TAG} from "./constants";
import makeSelectNavigationContainer, {selectNavigationContainerDomain} from "./selectors";
import {START_ADD} from "../ArticlesListContainer/constants";

// Individual exports for testing
export function* getTags() {
    try {
        const tags = yield call(fetchTags);
        yield put(getTagsSuccess(tags));
    }
    catch (error) {
        yield put(getTagsFailed(error.message));
    }
}

export function* watchTagsLoad() {
    yield takeLatest(GET_TAGS, getTags);
}

export function* setDefaultArticles(action) {
    const state = yield select(makeSelectNavigationContainer());
    console.log(state);
    if(Object.entries(state.selectedTag).length === 0 && state.routerLocation === '/'){
        // todo display all articles
        yield put(push('/topics'))
    }
}

export function* watchDefaultArticlesSetting() {
    yield takeLatest(GET_TAGS_SUCCESS, setDefaultArticles);
}

export function* pushTag(action){
    yield put(push(`/topics/${action.selectedTag.name}`))
}

export function* watchTagsPush() {
    yield takeLatest(SELECT_TAG, pushTag);
}

export function* startArticleAdd(action) {
    yield put(push(`/topics/${action.tagName.name}/add`));
}

export function* watchArticleAdd() {
    yield takeLatest(START_ADD, startArticleAdd);
}

export default function* rootSaga() {
    yield all([
        watchTagsLoad(),
        watchTagsPush(),
        watchDefaultArticlesSetting(),
        watchArticleAdd(),
    ])
}
