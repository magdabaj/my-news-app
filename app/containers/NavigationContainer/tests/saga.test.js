/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { take, call, put, takeLatest } from 'redux-saga/effects';
import { push } from 'react-router-redux';
// import navigationContainerSaga from '../saga';

// const generator = navigationContainerSaga();

import {getTags, pushTag, pushTagTest, watchTagsLoad, watchTagsPush} from "../saga";
import {getTagsFailed, getTagsSuccess} from "../actions";
import {GET_TAGS, SELECT_TAG} from "../constants";

describe("navigationContainerSaga Saga", () => {
  let getTagsGenerator;

  beforeEach(() => {
    getTagsGenerator = getTags();

    const callDescriptor = getTagsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot()
  });

  it("should dispatch getTagsSuccess if it gets the data successfully", () => {
    const response = {
      tags: [
        {
          id: 1,
          name: 'saga'
        },
        {
          id: 2,
          name: 'redux'
        }
      ]
    }
    const putReceivedTagsSuccess = getTagsGenerator.next(response).value;
    expect(putReceivedTagsSuccess).toEqual(
        put(getTagsSuccess(response))
    );
  });

  it('should call the getTagsFailed action if the response errors', () => {
    const response = new Error('Some error')
    const putReceivedTagsFailed = getTagsGenerator.throw(response).value
    expect(putReceivedTagsFailed).toEqual(
        put(getTagsFailed(response.message))
    )
  })
});

describe('loading tags saga', () => {
  const loadTagsSaga = watchTagsLoad();

  it('should start a task to watch for GET_TAGS action', () => {
    const takeLatestDescriptor = loadTagsSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
        takeLatest(GET_TAGS, getTags)
    )
  })
})

describe('push tag saga', () => {
  let pushTagGenerator;

  const action = {
    type: SELECT_TAG,
    selectedTag: {
      id: 1,
      name: 'saga'
    }
  };
  beforeEach(() => {
    pushTagGenerator = pushTag(action);

    const callDescriptor = pushTagGenerator.next(action.selectedTag.name).value;
    expect(callDescriptor).toMatchSnapshot();
  });
})

describe('watch tag push', () => {
  const watchTagsPushSaga = watchTagsPush();

  it('should start task to watch SELECT_TAG action', () => {
    const takeLatestDescriptor = watchTagsPushSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(SELECT_TAG, pushTag))
  })
})
