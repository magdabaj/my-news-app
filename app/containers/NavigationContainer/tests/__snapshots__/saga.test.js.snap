// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigationContainerSaga Saga should call the getTagsFailed action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`navigationContainerSaga Saga should dispatch getTagsSuccess if it gets the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
